### PR TITLE
LOK-2233 : BFF: Node: SearchIpInterfaces

### PR DESCRIPTION
**Pull request by Junaid-Khan-Nant**
## Description
1. Add gRPC service in nodeService.proto to search IpInterfaces either by IpAddress or by hostName and return list of IpInterface ( IpInterfaceDTO in ipInterface.proto.
2. Implement this in Inventory service along with a test case.
3. Use this new service in BFF ( rest-server) to implement new GraphQL service that will take a string and return List of IpInterfaces.

## Jira link(s)
https://opennms.atlassian.net/browse/LOK-2233

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

**Review by smunir-onms — COMMENTED** [review]
looks good to me

File: shared-lib/lokahi-common-utils/src/main/java/org/opennms/horizon/shared/utils/InetAddressUtils.java [hunk 1]
```diff
@@ -636,4 +636,13 @@ public static String getHumanReadableIfSpeed(long ifSpeed) {
 
         return formatter.format(displaySpeed) + " " + units;
     }
+
+    public static boolean isValidIpAddress(String ipAddress) {
```
Comment (smunir-onms) [inline shared-lib/lokahi-common-utils/src/main/java/org/opennms/horizon/shared/utils/InetAddressUtils.java:640]: minor: name should be more descriptive: isValidIpv4OrIpv6Address

**Review by smunir-onms — COMMENTED** [review]
LGTM

File: shared-lib/lokahi-common-utils/src/main/java/org/opennms/horizon/shared/utils/InetAddressUtils.java (hunk 1 above)
Comment (Junaid-Khan-Nant) [reply shared-lib/lokahi-common-utils/src/main/java/org/opennms/horizon/shared/utils/InetAddressUtils.java]: Done

File: inventory/main/src/main/java/org/opennms/horizon/inventory/repository/IpInterfaceRepository.java [hunk 2]
```diff
@@ -63,4 +63,15 @@ Optional<IpInterface> findByIpLocationIdTenantAndScanType(
...
+
+    @Query("SELECT ip " + "FROM IpInterface ip "
+            + "WHERE ip.tenantId = :tenantId "
+            + "AND ip.node.id = :nodeId "
+            + "AND (ip.ipAddress = :ipAddress "
+            + "OR LOWER(ip.hostname) LIKE LOWER(CONCAT('%', :searchTerm, '%')))")
+    List<IpInterface> findBytenantIdAndSearchIpInterfacesTerm(
+            @Param("tenantId") String tenantId,
+            @Param("nodeId") long nodeId,
+            @Param("ipAddress") InetAddress ipAddress,
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/repository/IpInterfaceRepository.java:75]: A node has multiple IpInterfaces and each IpInterfaces is an IpAddress ( InetAddress), you have to search in all of those IpAdresses not one IpAddress

**Review by cgorantla — CHANGES_REQUESTED** [review]

File: shared-lib/inventory/src/main/proto/nodeService.proto [hunk 3]
```diff
@@ -118,4 +123,5 @@ service NodeService {
   rpc getIpInterfaceById(google.protobuf.Int64Value) returns (IpInterfaceDTO) {};
   rpc getIpInterfaceFromQuery(NodeIdQuery) returns (IpInterfaceDTO) {};
   rpc getNodeCount(google.protobuf.Empty) returns (google.protobuf.Int64Value) {};
+  rpc listSearchIpInterfaceByQuery(SearchIpInterfaceQuery) returns (IpInterfaceList) {};
```
[the remark this answers is not in the document]
Comment (Junaid-Khan-Nant) [reply shared-lib/inventory/src/main/proto/nodeService.proto]: Done

File: inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java [hunk 4]
```diff
@@ -384,4 +387,22 @@ public List<NodeDTO> listNodesByTags(String tenantId, List<String> tags) {
                 .map(mapper::modelToDTO)
                 .toList();
     }
+
+    @Transactional(readOnly = true)
+    public List<IpInterfaceDTO> listSearchIpInterfacesByQuery(
+            String tenantId, Long nodeId, String searchIpInterfaceTerm) {
+        List<IpInterface> ipInterfaces = null;
+        InetAddress addr = null;
+        if (InetAddressUtils.isValidIpv4OrIpv6Address(searchIpInterfaceTerm)) {
```
[the remark this answers is not in the document]
Comment (Junaid-Khan-Nant) [reply inventory/main/src/main/java/org/opennms/horizon/inventory/service/NodeService.java]: Done

**Review by smunir-onms — COMMENTED** [review]
SGTM

**Review by smunir-onms — COMMENTED** [review]
updated changes seems OK

File: shared-lib/lokahi-common-utils/src/main/java/org/opennms/horizon/shared/utils/InetAddressUtils.java [hunk 5]
```diff
@@ -636,4 +636,13 @@ public static String getHumanReadableIfSpeed(long ifSpeed) {
 
         return formatter.format(displaySpeed) + " " + units;
     }
+
+    public static boolean isValidIpv4OrIpv6Address(String ipAddress) {
```
Comment (cgorantla) [inline shared-lib/lokahi-common-utils/src/main/java/org/opennms/horizon/shared/utils/InetAddressUtils.java:640]: unused, remove

File: inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java [hunk 6]
```diff
@@ -479,4 +481,35 @@ private void sendScannerTasksToMinion(Map<Long, List<NodeDTO>> locationNodes, St
...
+                tenantId -> {
+                    try {
+                        List<IpInterfaceDTO> ipInterfaceList = nodeService.listSearchIpInterfacesByQuery(
+                                tenantId, request.getNodeId(), request.getSearchTerm());
+                        responseObserver.onNext(IpInterfaceList.newBuilder()
+                                .addAllIpInterface(ipInterfaceList)
+                                .build());
+                        responseObserver.onCompleted();
+                    } catch (Exception e) {
+
```
Comment (cgorantla) [inline inventory/main/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java:499]: log the exception

File: inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java [hunk 7]
```diff
@@ -303,4 +306,47 @@ public void fetchAListOfNodesByNodeNode_aliasWithSearchTerm(String aliasSearchTe
...
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("Node " + searchTerm + " not found"));
+
+        ipInterfaceList = nodeServiceBlockingStub.searchIpInterfaces(SearchIpInterfaceQuery.newBuilder()
+                .setNodeId(node.getId())
+                .setSearchTerm(searchTerm)
+                .build());
+    }
+
+    @Then("verify the list of IpInterfaces has size greater than {int}")
```
Comment (cgorantla) [inline inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java:348]: you can combine these two steps

**Review by cgorantla — CHANGES_REQUESTED** [review]

**Review by smunir-onms — COMMENTED** [review]
requested changes verified

File: inventory/docker-it/src/test/resources/org/opennms/horizon/inventory/node.feature [hunk 8]
```diff
@@ -35,3 +35,8 @@ Feature: Node
   Scenario: Add a node
     Given a new node with node_alias "node", label "label", ip address "127.0.0.1" in location named "Default"
     Then fetch the list of nodes response not equal to 0
+
+  Scenario: Search IpInterfaces by Node
+    Given a new node with IpInterface along with node label "my-label" ip address "128.0.0.1" in location named "Default"
+    Then verify that a new node is created with the ip address "128.0.0.1"
+    Then fetch a list of IpInterfaces by node using search term "128.0.0.1" that has size greater than 0
```
Comment (cgorantla) [inline inventory/docker-it/src/test/resources/org/opennms/horizon/inventory/node.feature:42]: update this  `Then fetch a list of IpInterfaces by node using search term "my" that has size greater than 0`. and see if it passes

File: inventory/docker-it/src/test/resources/org/opennms/horizon/inventory/inventory-processing.feature [hunk 9]
```diff
@@ -76,5 +76,8 @@ Feature: Inventory Processing
     Given Node Scan results with IpInterfaces "192.168.1.45" and SnmpInterfaces with ifName "eth0"
     Then  Send node scan results to kafka topic "task-set.results"
     Then verify node has IpInterface "192.168.1.45" and SnmpInterface with ifName "eth0"
-
+    Given Node Scan results with IpInterfaces "192.168.1.48" and hostName "Local"
+    Then Send node scan results to kafka topic for hostName and ipAddress "task-set.results"
+    Then verify node has IpInterface with hostName "Local"
```
Comment (cgorantla) [inline inventory/docker-it/src/test/resources/org/opennms/horizon/inventory/inventory-processing.feature:81]:
```suggestion
    Then verify node has IpInterface with hostName "Loc"
```
update this to partial search and see if it works

File: inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java [hunk 10]
```diff
@@ -303,4 +306,44 @@ public void fetchAListOfNodesByNodeNode_aliasWithSearchTerm(String aliasSearchTe
...
+                .build());
+    }
+
+    @Then("verify that a new node is created with the ip address {string}")
+    public void verifyThatANewNodeIsCreatedWithTheIpAddress(String ipAddress) {
+        var nodeServiceBlockingStub = backgroundHelper.getNodeServiceBlockingStub();
+        NodeDTO node = nodeServiceBlockingStub.listNodes(Empty.getDefaultInstance()).getNodesList().stream()
+                .filter(fetched -> ipAddress.equals(fetched.getIpInterfaces(0).getIpAddress()))
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("Node " + ipAddress + " not found"));
```
Comment (cgorantla) [inline inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java:327]: don't think we need this check

**Review by cgorantla — COMMENTED** [review]
few comments otherwise looks good to me.

File: inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java (hunk 10 above)
Comment (Junaid-Khan-Nant) [reply inventory/docker-it/src/test/java/org/opennms/horizon/inventory/cucumber/steps/NodeStepDefinitions.java]: Removed.

File: inventory/docker-it/src/test/resources/org/opennms/horizon/inventory/inventory-processing.feature (hunk 9 above)
Comment (Junaid-Khan-Nant) [reply inventory/docker-it/src/test/resources/org/opennms/horizon/inventory/inventory-processing.feature]: Its working with partial search.

File: inventory/docker-it/src/test/resources/org/opennms/horizon/inventory/node.feature (hunk 8 above)
Comment (Junaid-Khan-Nant) [reply inventory/docker-it/src/test/resources/org/opennms/horizon/inventory/node.feature:41]: As discuss its is done.

**Review by smunir-onms — COMMENTED** [review]
latest changes verified.

**Review by cgorantla — APPROVED** [review]
LGTM!

**Comment by Junaid-Khan-Nant** [comment]
Merge LOK-2233 into develop.

**Pull request merged**